Add recommendations around reducing big bang release risk (#257)
18F/development-guide: _pages/release-strategies.md

---
title: Releasing Software
sidenav: approach
sticky_sidenav: true
---

Releasing software is hard. Over time, we have developed strategies that can help reduce risk and uncertainity in the release process.

## Small, Iterative Releases

We recommend on all projects to get into the habit of releasing software as early and often as possible. The first
iteration of your software should not aim to have full functionality; a better goal for the very first release is to have as little
functionality as possible.

## Big Bang Releases

We do not recommend "big bang releases," defined here as when a team works many months on some code with the expectation that it will be

> "turned on" in a big release event. 

Author: i want to create an issue that talks about this more since i think the why behind this is realllllly important. but i think a placeholder link is a great idea

> "turned on" in a big release event. 

Reviewer: Prior stuff: https://www.google.com/search?q=18f+iterative+release&oq=18f+iterative+release

#### Develop Training Strategies in Advance

Recommendation: Do user research and develop training strategies ahead of time to help users ease into the new system.  

It can be jarring as a user to one day have a completely new system. Spending some time before release working with critical users of the
system to understand the common pitfalls a user may experience or uncover will help you develop training materials to address those issues 
(or make the system more intuitive.) Releasing training materials for users to look at before the release will create better familiarity and
make users less wary or uncomfortable with the big change.

#### Prepare Customer Support Response Templates

Recommendation: Develop templates for expected support requests to have consistent messaging and advice to users.

Launches often have expected support requests, such as: 
* "Why doesn't the system do X?"
* "This is broken."
* "I can't find Z."

Preparing responses ahead of time allows the customer support team to give the same suppport to anyone who a predictable issue, allowing them
to focus their energies on unexpected support needs. It also helps unsure a better customer experience for users.

> ### Big Bang Release Factors

Reviewer: Another thing we might want to address is that most of these are project smells, too. For example, if they won't pick a set of users, then they can't do usability testing, which in turn is a big red flag for the project in general. If you're seeing these, you need to be rolling them up to your project health tracking.

There are some factors that make partners more attached to big bang releases. There are some strategies you can deploy to try to 

> convince partners to pivot to a smaller, more iterative release strategy.

Reviewer: I would suggest the number one thing you can do  -- if you can manage it -- is getting the partner to pivot from a project-centric to a product-centric approach, which pays off in about 100 different ways. In the former, the partner sees the big bang release as the culmination and end point of the project, which is *really* problematic and one of the biggest issues gov IT faces. In the latter, they understand that the release is just one milestone in the lifecycle. I know we've written a lot about that project-to-product shift in the past (and I'm sure the product chapter would happily weigh in). At least IMO, that's the biggest and important nut to crack.

Reviewer: I think we've almost definitely written about it many times before, so maybe we can reference and link?

Author: i did a quick search of the 18f blog and couldn't find it specifically; do you have a link you could pass along?

Reviewer: I wonder if this series would work well to link to? https://18f.gsa.gov/2019/08/22/building-product-management-capacity-in-government-part-1/

> #### Partner Doesn't Want to Pick a Cohort of Controlled Users

Reviewer: ```suggestion
#### The partner doesn't want to pick a cohort of controlled users
```

Understanding why a partner doesn't want to single out some controlled users is important. They may be nervous about creating inequity (or

> a perception of inequity) that might have some polical blowback. Or perhaps they feel overwhelmed about the effort required to narrow down

Reviewer: ```suggestion
a perception of inequity) that might have some political blowback. Perhaps they feel overwhelmed about the effort required to narrow down
```

to a subset of users, and so feel it's simpler to release to all users at once.

> Mitigation strategies:

Reviewer 2: These are formatted on separate lines in the markdown, but not marked up as an unordered list. Do we want to use a `* ` for each of them, or else join the lines in the markdown to more closely reflect that they are going to be paragraph form in the rendered version? (same for the next two sections)

Author: I wasn't going to make them an unordered list, but on reflection I think it's more readable that way! Will update; thanks for the suggestion!

> * Work with your partner to fully understand their fears. 

Reviewer: ```suggestion
* Talk with your partner to fully understand their fears. 
```

Reviewer: You're not trying to work the problem here, you're trying to actively listen to understand. So I'm not sure "talk" is the verb I want either, but it's closer.

* Frame this step as the first of many interactions with their different user groups, and use that framing to help them better engage and 

> understand their users. Having a more mature idea of their user groups will help them have confidence to stand behind their decisions better.

Reviewer: Our user researchers explicitly recommend working with different types of users for better testing and a fuller picture. Maybe we could better bridge that, tying research goals to the stakeholder concerns and show how both want a diverse user group?

> #### A Long, Arduous ATO Process

Reviewer: ```suggestion
#### A long, arduous ATO process
```

Partners have a valid fear of the ATO process, and may be nervous that they must repeat the ATO process every time a change is made to the

> system. That perspective easily lends itself to fear of iterative releases. 

Reviewer: I can't recall where, but we've written a lot before about modular ATOs or iterative ATOs and such. It might be worth digging those up.

*Mitigation strategies:*

> * Engage security or ATO personnel early in the development process, or, ideally, embed someone onto the project team who can help advise.

Reviewer: ```suggestion
* Engage security and ATO personnel early in the development process, or, ideally, embed someone onto the project team who can help advise.
```

Reviewer: Suggestion:

* Familiarize yourself with 18Fs "Before you ship" guide so you can personally help alleviate some of their concerns.

* Teams at 18F have also found success in employing the "Walking Skeleton" technique, where the main architectural components of a system are
deployed early in a minumum viable way. Frontloading the infrastructure work creates an MVP for ATO work, and makes space for early
compliance and security oversight.

> #### Legislative Mandate

Reviewer: ```suggestion
#### Legislative mandate
```
